src/internet_identity/tests/tests.rs

```diff
@@ -497,6 +466,108 @@ mod registration_tests {
         );
         Ok(())
     }
+
+    /// Tests that the `register` call will hit the rate limit on too many calls and that the limit
+    /// will allow new calls after some time.
+    #[test]
+    fn should_rate_limit_register_calls() -> Result<(), CallError> {
+        let env = env();
+        let canister_id = install_ii_canister_with_arg(
+            &env,
+            II_WASM.clone(),
+            arg_with_rate_limit(RateLimitConfig {
+                time_per_token_ns: Duration::from_secs(1).as_nanos() as u64,
+                max_tokens: 2,
+            }),
+        );
+
+        for _ in 0..2 {
+            flows::register_anchor(&env, canister_id);
+        }
+        let challenge = api::create_challenge(&env, canister_id)?;
+        let result = api::register(
+            &env,
+            canister_id,
+            principal_1(),
+            &device_data_1(),
+            ChallengeAttempt {
+                chars: "a".to_string(),
+                key: challenge.challenge_key.clone(),
+            },
+        );
+        expect_user_error_with_message(
+            result,
+            CanisterCalledTrap,
+            Regex::new("rate limit reached, try again later").unwrap(),
+        );
+        assert_metric(
+            &get_metrics(&env, canister_id),
+            "internet_identity_register_rate_limit_current_tokens",
+            0f64,
+        );
+
+        env.advance_time(Duration::from_secs(1));
+
+        let result = api::register(
+            &env,
+            canister_id,
+            principal_1(),
+            &device_data_1(),
+            ChallengeAttempt {
+                chars: "a".to_string(),
+                key: challenge.challenge_key,
+            },
+        );
+        assert!(result.is_ok());
+        Ok(())
+    }
+
+    /// Tests that the `register` rate limit does not replenish tokens to more than max_tokens.
+    #[test]
+    fn should_not_allow_more_than_max_tokens_calls_on_rate_limit() -> Result<(), CallError> {
+        let env = env();
+        let canister_id = install_ii_canister_with_arg(
+            &env,
+            II_WASM.clone(),
+            arg_with_rate_limit(RateLimitConfig {
+                time_per_token_ns: Duration::from_secs(1).as_nanos() as u64,
+                max_tokens: 2,
+            }),
+        );
+
+        env.advance_time(Duration::from_secs(100));
+
+        // some activity required to process the rate limit (which will use one token)
+        flows::register_anchor(&env, canister_id);
+
+        assert_metric(
+            &get_metrics(&env, canister_id),
+            "internet_identity_register_rate_limit_current_tokens",
+            1f64, // 1 and not 2 because of register call required to update rate limit
```

Comment:
nit: maybe use a more realistic value for `max_tokens`?

Reply:
Nvm, as discussed added higher limit.